File: Project2/Solutions/python/tic_tac_toe.py
""" Part II - Problem 1 (Tic-Tac-Toe) """

import sys
import random
import math
import copy



class TicTacToe:
    """ Tic-Tac-Toe Game Class """
    def __init__(self):
        self.board = [0] * 9
        self.xs_turn = True
        self.output_str = None

    def play(self):
        """ High level function call that starts the game """
        while not self.terminal_test(self.board):
            self.move()
            self.add_to_output_string()
        self.print_boards_to_file()

    def move(self):
        """ high level move function, selects which player to move """
        if self.xs_turn:
            self.move_x()
        else:
            self.move_o()
        self.xs_turn = not self.xs_turn

    def move_x(self):
        """ move x randomly """
        space = self.rand_space()
        while self.board[space] is not 0:
            space = self.rand_space()
        self.board[space] = 1

    def move_o(self):
        """ move o using minimax algorithm """
        action = self.minimax_decision(self.board)
        self.board[action[0]] = action[1]

    def minimax_decision(self, board, player=2):
        """ minimax decision algorithm """
        action_set = self.get_action_set(board, player)

        max_index = 0
        max_util = -math.inf
        for idx, action in enumerate(action_set):
            curr_util = self.min_value(self.result(board, action))
            if curr_util > max_util:
                max_util = curr_util
                max_index = idx

        return action_set[max_index]

    def max_value(self, board):
        """ max_value function from minimax algorithm """
        if self.terminal_test(board):
            return self.utility(board)

        max_util = -math.inf
        action_set = self.get_action_set(board, 2)

        for action in action_set:
            new_util = self.min_value(self.result(board, action))
            max_util = max(max_util, new_util)

        return max_util

    def min_value(self, board):
        """ min_value function from minimax algorithm """
        if self.terminal_test(board):
            return self.utility(board)

        min_util = math.inf
        action_set = self.get_action_set(board, 1)

        for action in action_set:
            new_util = self.max_value(self.result(board, action))
            min_util = min(min_util, new_util)

        return min_util

    def result(self, board, action):
        """ obtains new copy of board with action taken """
        new_board = copy.deepcopy(board)
        new_board[action[0]] = action[1]
        return new_board

    def terminal_test(self, board):
        """ Returns false if someone has won OR if there are no moves to play, true otherwise """
        return self.win_check(board, 1) or self.win_check(board, 2) or self.no_moves_left(board)

    def no_moves_left(self, board):
        """ Returns True if there are no moves left to play """
        ret_val = all(space != 0 for space in board)
        # print("No moves left = ",ret_val)
        return ret_val

    def win_check(self, board, p):
        """ Returns True if player p has won, false otherwise """
        # Column wins
        col1 = (board[6] == p) and (board[3] == p) and (board[0] == p)
        col2 = (board[7] == p) and (board[4] == p) and (board[1] == p)
        col3 = (board[8] == p) and (board[5] == p) and (board[2] == p)

        # Row wins
        row1 = (board[6] == p) and (board[7] == p) and (board[8] == p)
        row2 = (board[3] == p) and (board[4] == p) and (board[5] == p)
        row3 = (board[0] == p) and (board[1] == p) and (board[2] == p)

        # Diagonal wins
        diag1 = (board[0] == p) and (board[4] == p) and (board[8] == p)
        diag2 = (board[2] == p) and (board[4] == p) and (board[6] == p)

        ret_val = (col1 or col2 or col3 or row1 or row2 or row3 or diag1 or diag2)
        # print("Win check = ",ret_val)
        return ret_val


    def utility(self, board):
        """ Returns the utility value of the board """
        if self.win_check(board, 2):
            return 1.0
        if self.win_check(board, 1):
            return 0.0
        return 0.5

    def get_action_set(self, board, player):
        """ Obtain a list of the available actions """
        action_set = []
        for i in range(0, 9):
            if board[i] == 0:
                new_action = (i, player)
                action_set.append(new_action)
        return action_set

    def rand_space(self):
        """ Generate a random integer in [0,8] """
        return math.floor(9 * random.random())

    def add_to_output_string(self):
        """Add current board to output string """
        # 1) Pre-process values to characters
        chars = ['-'] * 9
        for idx, val in enumerate(self.board):
            chars[idx] = self.val_to_char(val)

        # 2) Add the characters to the output string
        if self.output_str is None:
            self.output_str = ""
        else:
            self.output_str += "\n\n"
        self.output_str += chars[6] + "," + chars[7] + "," + chars[8] + "\n"
        self.output_str += chars[3] + "," + chars[4] + "," + chars[5] + "\n"
        self.output_str += chars[0] + "," + chars[1] + "," + chars[2]


    def print_boards_to_file(self):
        """ Print the output string to the file """
        f = open("tic_tac_toe.txt", "w")
        f.write(self.output_str)
        f.close()

    def val_to_char(self, val):
        """ convenience function for printing """
        if val == 0:
            return '-'
        if val == 1:
            return 'x'
        return 'o'


def main(prng_seed=42):
    """ main function to call tic_tac_toe code """

    # 1) Seed the PRNG
    random.seed(prng_seed)

    # 2) Init the Tic-Tac-Toe Game
    my_tic_tac_toe = TicTacToe()

    # 3) Play the game
    my_tic_tac_toe.play()




def cli_help():
    """ Command Line Interface Help Function """
    print("Please call tic_tac_toe.py as follows: python tic_tac_toe.py <seed> ")
    print("<seed> = (any number to initialize the PRNG)),")
    print()
    print("Example - : python tic_tac_toe.py 42")


if __name__ == "__main__":
    if len(sys.argv) is not 2:
        cli_help()
    else:
        main(sys.argv[1])

File: Project2/Solutions/cpp/tic_tac_toe.cpp
#include "tic_tac_toe.h"
using std::cout;
using std::endl;
using std::vector;


//////////////////////////////////////////////////////////////////////////////
//                                                                          //
//                                                                          //
//                         Gameplay Functions                               //
//                                                                          //
//                                                                          //
//////////////////////////////////////////////////////////////////////////////

/*
*   Plays tic tac toe
*/
void ticTacToe::play() {
     // Init the board
    vector<int> board(9, 0);

    while (boardValid(board)) {
        move(&board);
        // printBoardToTerminal(board);
        printBoardToFile(board);
    }
}

/*
*   moves either x or o depending on who's turn it is
*
*   And then change who's turn it is
*/
void ticTacToe::move(vector<int>* board) {
    if (Xs_Turn) {
        moveX(board);
    } else {
        moveO(board);
    }

    Xs_Turn = !Xs_Turn;
}

/*
*   Randomly sample spaces until a free one is found
*
*   Then, choose that as the move
*/
void ticTacToe::moveX(vector<int>* board) {
    int space = randSpace();
    while ((*board)[space] != 0) {
        space = randSpace();
    }

    (*board)[space] = 1;
}

/*
*   Uses the Minimax algorithm to choose a move
*/
void ticTacToe::moveO(vector<int>* board) {
    Action nextAction = minimaxDecision((*board), 2);
    (*board)[nextAction.space] = nextAction.player;
}




//////////////////////////////////////////////////////////////////////////////
//                                                                          //
//                                                                          //
//                         Minimax Functions                                //
//                                                                          //
//                                                                          //
//////////////////////////////////////////////////////////////////////////////


/*
*   Minimax Algorithm
*
*   Returns an action 
*/
Action ticTacToe::minimaxDecision(State state, int player) {
    // Obtain the action set
    vector<Action> actionSet = getActionSet(state, player);

    // Compute arg max over the action set of minValue of each action
    int maxIndex = 0;
    float maxUtil = negInf;
    for (unsigned int i=0; i < actionSet.size(); i++) {
        float currUtil = minValue(result(state, actionSet[i]));
        if (currUtil > maxUtil) {
            maxUtil = currUtil;
            maxIndex = i;
        }
    }

    return actionSet[maxIndex];
}


/*
*   maxValue
*
*   Returns a utility value
*/
float ticTacToe::maxValue(State state) {
    if (terminalTest(state)) {
        float util = utility(state);
        return util;
    }

    float v = negInf;
    vector<Action> actionSet = getActionSet(state, 2);

    // Compute the max utility of the min value from the action set
    for (unsigned int i=0; i < actionSet.size(); i++) {
        float newV = minValue(result(state, actionSet[i]));
        if (newV > v) {
            v = newV;
        }
    }

    return v;
}

/*
*   minValue
*
*   In our case player 1 (x) is expected to be going
*
*   Returns a utility value
*/
float ticTacToe::minValue(State state) {
    if (terminalTest(state)) {
        float util = utility(state);
        return util;
    }

    float v = posInf;
    vector<Action> actionSet = getActionSet(state, 1);

    // Compute the min utility of the max value from the action set
    for (unsigned int i=0; i < actionSet.size(); i++) {
        float newV = maxValue(result(state, actionSet[i]));
        if (newV < v) {
            v = newV;
        }
    }

    return v;
}

/*
*   getActionSet
*
*   Returns a vector of available Actions for the player 
*/
vector<Action> ticTacToe::getActionSet(State state, int player) {
    vector<Action> actionSet;

    for (int i = 0; i < 9; i++) {
        if (state[i] == 0) {
            Action newAction;
            newAction.player = player;
            newAction.space = i;
            actionSet.push_back(newAction);
        }
    }

    return actionSet;
}

/*
*   terminalTest
*
*   Returns true if the game is over
*/
bool ticTacToe::terminalTest(State state) {
    bool bValid = boardValid(state);
    return !bValid;
}

/*
*   utility
*
*   Returns the utlity value of the given state (w.r.t. player 2 as Max)
*/
float ticTacToe::utility(State state) {
    float util;

    if (winCheck(state, 2)) {
        util = 1.0;
    } else if (winCheck(state, 1)) {
        util = 0.0;
    } else {
        util = 0.5;
    }

    // if neither player won, it's a cats game!
    return util;
}


//////////////////////////////////////////////////////////////////////////////
//                                                                          //
//                                                                          //
//                         Helper Functions                                 //
//                                                                          //
//                                                                          //
//////////////////////////////////////////////////////////////////////////////


/*
*   Returns a random int from 0-8
*/
int ticTacToe::randSpace() {
    return floor(9*((double) rand())/RAND_MAX);
}


void ticTacToe::help() {
    cout << "Usage: " << endl << endl;

    cout << "   ./tic_tac_toe <seed> " << endl << endl;

    cout << "   where <seed> is an integer # used to seed the PRNG";
    cout << endl << endl;
}

void ticTacToe::printBoardToTerminal(vector<int> board) {
    cout << endl;
    cout << v2C(board[6]) <<","<< v2C(board[7]) <<","<< v2C(board[8]) << endl;
    cout << v2C(board[3]) <<","<< v2C(board[4]) <<","<< v2C(board[5]) << endl;
    cout << v2C(board[0]) <<","<< v2C(board[1]) <<","<< v2C(board[2]) << endl;
}

void ticTacToe::printBoardToFile(vector<int> board) {
    if (!myfile.is_open()) {
        myfile.open("tictactoe.txt");
    }

    if (!first_board_printed) {
        first_board_printed = true;
    } else {
        myfile << endl << endl;
    }

    myfile << v2C(board[6]) <<","<< v2C(board[7]) <<","<< v2C(board[8]) << endl;
    myfile << v2C(board[3]) <<","<< v2C(board[4]) <<","<< v2C(board[5]) << endl;
    myfile << v2C(board[0]) <<","<< v2C(board[1]) <<","<< v2C(board[2]);
}


/*
*   Value 2 Char
*/
char ticTacToe::v2C(int val) {
    switch (val) {
        case 0:
            return '-';
            break;
        case 1:
            return 'x';
            break;
        case 2:
            return 'o';
            break;
        default:
            return '-';
    }
}


/*
*   Super verbose wincheck for player p
*       Could make it faster but for now who cares...
*/
bool ticTacToe::winCheck(vector<int> board, int p) {
    // Column wins
    bool col1 = (board[6] == p) && (board[3] == p) && (board[0] == p);
    bool col2 = (board[7] == p) && (board[4] == p) && (board[1] == p);
    bool col3 = (board[8] == p) && (board[5] == p) && (board[2] == p);

    // Row wins
    bool row1 = (board[6] == p) && (board[7] == p) && (board[8] == p);
    bool row2 = (board[3] == p) && (board[4] == p) && (board[5] == p);
    bool row3 = (board[0] == p) && (board[1] == p) && (board[2] == p);

    // Diagonal wins
    bool diag1 = (board[0] == p) && (board[4] == p) && (board[8] == p);
    bool diag2 = (board[2] == p) && (board[4] == p) && (board[6] == p);

    return (col1 || col2 || col3 || row1 || row2 || row3 || diag1 || diag2);
}

/*
*   Again a super verbose no moves checker
*       could make it faster...
*/
bool ticTacToe::noMovesLeft(vector<int> board) {
    return ((board[6] != 0) && (board[7] != 0) && (board[8] != 0) &&
            (board[3] != 0) && (board[4] != 0) && (board[5] != 0) &&
            (board[0] != 0) && (board[1] != 0) && (board[2] != 0) );
}


/*
*   Executes an action (creating a new state)
*/
vector<int> ticTacToe::result(vector<int> board, Action action) {
    board[action.space] = action.player;
    return board;
}


/*
*   Returns false if someone has won OR if there are no moves to play
*
*   Returns true otherwise
*/
bool ticTacToe::boardValid(vector<int> board) {
    if (winCheck(board, 1) || winCheck(board, 2)) {
        return false;
    }

    if (noMovesLeft(board)) {
        return false;
    }

    return true;
}



































